File: problems/57-insert-interval/solution.ts
function insertInterval(intervals: number[][], newInterval: number[]): number[][] {
    if (intervals.length < 1) return [newInterval];

    let insertS = false, insertE = false;
    let start = [], end = [];

    for (let i = 0; i < intervals.length; i++) {
        if (intervals[i][0] <= newInterval[0]) start.push(intervals[i])
        if (intervals[i][1] > newInterval[1]) end.push(intervals[i])
    }

    if (start.length > 0) {
        if (newInterval[0] <= start.at(-1)![1]) {
            if (start.at(-1)![1] < newInterval[1]) start.at(-1)![1] = newInterval[1]

            insertS = true;
        }
    }
    if (end.length > 0) {
        if (newInterval[1] >= end[0][0]) {
            if (end[0][0] > newInterval[0]) end[0][0] = newInterval[0]

            insertE = true;
        }
    }

    if (insertS && insertE) start.at(-1)![1] = end.shift()![1]
    else if (!(insertS || insertE)) start.push(newInterval);


    return start.concat(end);
};

export { insertInterval };

File: problems/57-insert-interval/solution.go
package main

func insertInterval(intervals [][]int, newInterval []int) [][]int {
	if len(intervals) < 1 {
		return [][]int{newInterval}
	}

	insertS, insertE := false, false
	start, end := [][]int{}, [][]int{}

	for _, iValue := range intervals {
		if iValue[0] <= newInterval[0] {
			start = append(start, []int{iValue[0], iValue[1]})
		}
		if iValue[1] > newInterval[1] {
			end = append(end, []int{iValue[0], iValue[1]})
		}
	}

	if len(start) > 0 {
		if newInterval[0] <= start[len(start)-1][1] {
			if start[len(start)-1][1] < newInterval[1] {
				start[len(start)-1][1] = newInterval[1]
			}

			insertS = true
		}
	}

	if len(end) > 0 {
		if newInterval[1] >= end[0][0] {
			if end[0][0] > newInterval[0] {
				end[0][0] = newInterval[0]
			}

			insertE = true
		}
	}

	if insertS && insertE {
		start[len(start)-1][1] = end[0][1]
		end = end[1:]
	} else if !(insertS || insertE) {
		start = append(start, []int{newInterval[0], newInterval[1]})
	}

	start = append(start, end...)
	return start
}
